WordPress/gutenberg
# Use git for `wp-env`'s default WordPress version

## What?
Up to now, the default WordPress version installed by `wp-env` is whatever Docker provides. This version is used when there is no core version specified in `wp-env.json`. We have no control over what Docker provides, and it is difficult to know what version it installs before it installs it. This causes problems when we wish to install a matching version of the phpunit test code, for example.

In this change, we now default to `WordPress/WordPress#$latest_version` instead, where `$latest_version` is pulled from https://api.wordpress.org/core/stable-check/1.0/.

There are two cons, but I think they are pretty minor:
- Default installs now have to spend a bit of extra time downloading WordPress with git. This should not have a huge impact, though. Perhaps 10 seconds longer running an update, which does not happen every time you start wp-env.
- Every run of `npx wp-env start` now includes an extra API request to `wordpress.org`. In my very rough testing, this only adds about 300ms at most.

## Why?
When trying to install a matching version of the phpunit test code, we need a WordPress version already available so that we can see which version is installed. However, Docker installs WordPress inside the container, which is difficult to read. There is also no guarantee it will always be up-to-date as we need it to be.

Controlling the WordPress version ourselves means that users will get WordPress updates more quickly. For example, when the WordPress update is published, `wp-env` will immediately start using it because our "cache" will immediately invalidate as the version in the configuration changes.

It also means that there will always be a `version.php` file for our "version-matcher" code to use for phpunit test code downloads.

For some extra context, I first noticed this issue trying to integrate the phpunit updates in `wp-env` in a different repository. I discovered that a filename was changed in core WordPress -- and this change was present in the test library, but not in WordPress itself. This is because the test code downloader uses `trunk` tests when no core source is given. But Docker installs the latest stable version (6.0.1) when no core source is given. So there was a mismatch.

This PR fixes that bug, since there is now a core source that our test code downloader can match in the default case.

## How?
1. Send a get request to https://api.wordpress.org/core/stable-check/1.0/, and parse the response for the latest version.
2. Set the default core source version to `WordPress/WordPress#...`, with a ref pointing at the latest version from step one. That repo has tags for every published WordPress version, matching the format from the stable check API.
3. The existing wp-env code will turn the string we set as the default core source into a normal git core source which gets processed as normal.

## Testing Instructions
1. Run `path/to/gutenberg/packages/env/bin/wp-env start` for a plugin which has no core source.
2. Verify that WordPress 6.0.1 is installed.
3. Verify that WordPress source code was pulled to `~/.wp-env/$hash/`

Also verify that there is no change for a repo like Gutenberg, which specifies a core source already.
